src/PPU.cpp

```diff
@@ -98,122 +69,273 @@ void PPU::updateLY() {
+	uint16_t secondHalfStart = 0x8800;
 
-	// If lcdc3 = 1, bg map starts at 0x9C00, otherwise 0x9800
-	uint16_t bg_start = lcdc3 ? 0x9C00 : 0x9800;
+	// If lcdc3 = 1, the background map starts at 0x9C00, otherwise 0x9800
+	bool lcdc3 = mmu->directRead(0xFF40) & (1 << 3);
+	uint16_t backgroundStart = lcdc3 ? 0x9C00 : 0x9800;
 
 	// If lcdc6 = 1, win map starts at 0x9C00, otherwise 0x9800
-	uint16_t win_start = lcdc6 ? 0x9C00 : 0x9800;
-
-	// Optimization: Use scaleX and scaleY to set the start offset, and then only add the actual screens pixels
-	// to pixelbuffer. Treat it like the screen rather than the background map
-	
-	auto set_line = [&](uint8_t x, uint8_t y, uint8_t hi, uint8_t lo) {
-		// Get palette index from leftmost to rightmost pixel
-		pixelBuffer[y * 256 + x    ] = palette[((hi >> 6) & (1 << 1)) | ((lo >> 7) & 1)];
-		pixelBuffer[y * 256 + x + 1] = palette[((hi >> 5) & (1 << 1)) | ((lo >> 6) & 1)];
-		pixelBuffer[y * 256 + x + 2] = palette[((hi >> 4) & (1 << 1)) | ((lo >> 5) & 1)];
-		pixelBuffer[y * 256 + x + 3] = palette[((hi >> 3) & (1 << 1)) | ((lo >> 4) & 1)];
-
-		pixelBuffer[y * 256 + x + 4] = palette[((hi >> 2) & (1 << 1)) | ((lo >> 3) & 1)];
-		pixelBuffer[y * 256 + x + 5] = palette[((hi >> 1) & (1 << 1)) | ((lo >> 2) & 1)];
-		pixelBuffer[y * 256 + x + 6] = palette[((hi >> 0) & (1 << 1)) | ((lo >> 1) & 1)];
-		pixelBuffer[y * 256 + x + 7] = palette[((hi << 1) & (1 << 1)) | ((lo >> 0) & 1)];
+	bool lcdc6 = mmu->directRead(0xFF40) & (1 << 6);
+	uint16_t windowStart = lcdc6 ? 0x9C00 : 0x9800;
+
+	// Update palette information
+	uint32_t bgpData = mmu->directRead(0xFF47);
+	uint32_t bgp[] = {
+		(bgpData & 0x03) >> 0,
+		(bgpData & 0x0C) >> 2,
+		(bgpData & 0x30) >> 4,
+		(bgpData & 0xC0) >> 6
 	};
 
+	// Iterate through 32x32 tiles for the background and window
 	for (int i = 0; i < 0x0400; i++) {
 		uint8_t x = i % 32;
 		uint8_t y = i / 32;
 
-		uint8_t index = mmu->read(bg_start + i);
-		uint16_t start = (index > 127) ? second_half_start : first_half_start;
-		
-		// TODO: clean up logic
-		// Read each pair of bytes and set each of the 8 lines of pixels
-		for (int j = 0; j < 8; j++) {
-			uint8_t lo = mmu->read(start + (index % 128) * 16 + j * 2);
-			uint8_t hi = mmu->read(start + (index % 128) * 16 + j * 2 + 1);
+		// Determine the index of the tile in the data, and determine the correct
+		// starting location for block of tile data given that index
+		uint8_t bi = mmu->directRead(backgroundStart + i);
+		uint16_t bs = (bi > 127) ? secondHalfStart : firstHalfStart;
+		bi %= 128;
 
-			set_line(x * 8, j + y * 8, hi, lo);
-		}
+		uint8_t wi = mmu->directRead(windowStart + i);
+		uint16_t ws = (wi > 127) ? secondHalfStart : firstHalfStart;
+		wi %= 128;
+
+		setTile(backgroundBuffer, MAP_WIDTH, bs, bi, x * 8, y * 8, bgp);
+		setTile(windowBuffer, MAP_WIDTH, ws, wi, x * 8, y * 8, bgp);
 	}
 }
 
-uint8_t PPU::getSCY() {
-	return mmu->directRead(0xFF42);
+void PPU::updateObjects() {
```

Comment:
I added a todo, but I think I want to revisit the objects map and calculate them on a per line basis to be more accurate.